File: examples/tlv++/src/query.cpp
// Addresses the Application Label directly by path, `6F/A5/50`, without
// walking the whole document by hand. See parse.cpp for the document itself.
#include <array>
#include <iomanip>
#include <iostream>

#include "tlv++/tlv.hpp"

// Same bytes as parse.cpp's document.
static const std::array<tlv::byte, 12> document = {
    tlv::byte(0x6F), tlv::byte(0x0A), tlv::byte(0x84), tlv::byte(0x03),
    tlv::byte(0x41), tlv::byte(0x42), tlv::byte(0x43), tlv::byte(0xA5),
    tlv::byte(0x03), tlv::byte(0x50), tlv::byte(0x01), tlv::byte(0x01)};

int main() {
    auto query = tlv::query::parse("6F/A5/50");
    if (!query) {
        std::cerr << "query parse error: " << query.error().message << "\n";
        return 1;
    }

    bool found = false;
    auto result = query->walk(
        tlv::bytes(document.data(), document.size()), tlv_reader_format_ber, tlv_ber_is_constructed,
        TLV_WALK_MAX_DEPTH, 16, [&found](const tlv::entry& entry, size_t /*depth*/, size_t offset) {
            std::cout << "6F/A5/50 = " << std::hex << std::uppercase << std::setw(2)
                      << std::setfill('0') << static_cast<int>(entry.value[0]) << std::dec
                      << " (offset " << offset << ")\n";
            found = entry.value.size() == 1 && entry.value[0] == tlv::byte(0x01);
            return TLV_VISIT_CONTINUE;
        });
    if (!result) {
        std::cerr << "query error: " << result.error().message << "\n";
        return 1;
    }
    return found ? 0 : 1;
}

File: bindings/python/opentlv/examples/query.py
"""Addresses elements of a Document by a `/`-separated path of hexadecimal
tags instead of walking node by node. See document.py for direct navigation
and parse.py for the document this query addresses.

Run with `python examples/query.py` from `bindings/python/opentlv`, after
installing both packages.
"""

import opentlv

# Same bytes as parse.py's document: an FCI Template (6F) holding a DF Name
# (84) and an FCI Proprietary Template (A5) holding an Application Label (50).
DOCUMENT = bytes([0x6F, 0x0A, 0x84, 0x03, 0x41, 0x42, 0x43, 0xA5, 0x03, 0x50, 0x01, 0x01])


def main() -> None:
    with opentlv.Document(DOCUMENT, opentlv.Format.BER) as document:
        # A multi-tag query walks straight to a deeply nested element,
        # without visiting 6F and A5 as separate steps.
        label = document.find_path("6F/A5/50")
        print(f"6F/A5/50 -> tag {label.tag} value {bytes(label.value).hex(' ').upper()}")

        # A one-tag query is the same as document.find() at the top level.
        assert document.find_path("6F") == document.find(opentlv.Tag(b"\x6f"))

        # No match is not an error: it is simply None, same as find().
        assert document.find_path("6F/99") is None

        # Malformed query text raises, distinct from "no match".
        try:
            document.find_path("6F//50")
        except opentlv.InvalidArgError as error:
            print(f"rejected: {error}")


if __name__ == "__main__":
    main()
